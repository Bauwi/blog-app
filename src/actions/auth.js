import { firebase, googleAuthProvider, githubAuthProvider } from '../firebase/firebase';

export const login = uid => ({
  type: 'LOGIN',
  uid
});

// select the login provider
export const startLogin = socialLogin => () => {
  const provider = socialLogin === 'google' ? googleAuthProvider : githubAuthProvider;
  return firebase.auth().signInWithPopup(provider);
};
export const logout = () => ({
  type: 'LOGOUT'
});

export const startLogout = () => () => firebase.auth().signOut();
