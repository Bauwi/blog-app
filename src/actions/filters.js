export const setTextFilter = (text = '') => ({
  type: 'SET_TEXT_FILTER',
  text
});

export const searchByKeyword = () => ({
  type: 'SEARCH_BY_KEYWORD'
});

export const searchByTitle = () => ({
  type: 'SEARCH_BY_TITLE'
});

export const searchByAuthor = () => ({
  type: 'SEARCH_BY_AUTHOR'
});

export const sortByDate = () => ({
  type: 'SORT_BY_DATE'
});

export const sortByStars = () => ({
  type: 'SORT_BY_STARS'
});

export const resetFilters = () => ({
  type: 'RESET_FILTERS'
});
