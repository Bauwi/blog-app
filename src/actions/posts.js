import database from '../firebase/firebase';
import { db } from '../firebase/firebase';

// Manage CRUD personnal posts
export const addPost = post => ({
  type: 'ADD_POST',
  post
});

export const startAddPost = (postData = {}) => (dispatch, getState) => {
  const { uid } = getState().auth;
  const {
    title = '',
    body = '',
    shortBody,
    keywords = {},
    createdAt = 0,
    stars = 0,
    author,
    authorId = uid
  } = postData;
  const post = {
    title,
    body,
    shortBody,
    keywords,
    createdAt,
    stars,
    author,
    authorId
  };
  return db
    .collection('posts')
    .add(post)
    .then((doc) => {
      dispatch(addPost({
        id: doc.id,
        ...post
      }));
    });
};

export const removePost = id => ({
  type: 'REMOVE_POST',
  id
});

export const startRemovePost = id => dispatch =>
  db
    .collection('posts')
    .doc(id)
    .delete()
    .then(() => {
      dispatch(removePost(id));
    });

export const editPost = (id, updates) => ({
  type: 'EDIT_POST',
  id,
  updates
});

export const startEditPost = (id, updates) => dispatch =>
  db
    .collection('posts')
    .doc(id)
    .update(updates)
    .then(() => {
      dispatch(editPost(id, updates));
    });

export const setPosts = posts => ({
  type: 'SET_POSTS',
  posts
});

export const startSetPosts = id => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('posts')
    .where('authorId', '==', `${uid}`)
    .get()
    .then((querySnapshot) => {
      const posts = [];
      querySnapshot.forEach((doc) => {
        posts.push({
          id: doc.id,
          ...doc.data()
        });
        console.log(doc.id, ' => ', doc.data());
        console.log(posts);
      });
      dispatch(setPosts(posts));
    });
};

// manage reading all posts
export const setPostsSample = posts => ({
  type: 'SET_POSTS_SAMPLE',
  posts
});

export const startSetPostsSample = sampleSize => dispatch =>
  db
    .collection('posts')
    .orderBy('createdAt', 'desc')
    .limit(50)
    .get()
    .then((querySnapshot) => {
      const posts = [];
      querySnapshot.forEach((doc) => {
        posts.push({
          id: doc.id,
          ...doc.data()
        });
      });
      dispatch(setPostsSample(posts));
    });
