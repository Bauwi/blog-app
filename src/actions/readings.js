import { db } from '../firebase/firebase';
import moment from 'moment';
import axios from 'axios';
import { startSetAuthorFromUserId } from './users';

// manage reading all posts
// set readings in home

export const readingssHasErrored = bool => ({
  type: 'READINGS_HAS_ERRORED',
  hasErrored: bool
});

export const readingsIsLoading = bool => ({
  type: 'READINGS_IS_LOADING',
  isLoading: bool
});

export const setPostsSample = posts => ({
  type: 'SET_POSTS_SAMPLE',
  posts
});

// Find the last 100 posts.
export const startSetPostsSample = () => (dispatch) => {
  dispatch(readingsIsLoading(true));
  return db
    .collection('posts')
    .orderBy('createdAt', 'desc')
    .limit(100)
    .get()
    .then((querySnapshot) => {
      const posts = [];
      querySnapshot.forEach((doc) => {
        posts.push({
          id: doc.id,
          ...doc.data()
        });
      });
      dispatch(readingsIsLoading(false));
      return posts;
    })
    .then(posts => dispatch(setPostsSample(posts)))
    .catch(() => dispatch(readingssHasErrored(true)));
};

// manage reading specific user posts.

export const setSpecificUserPosts = posts => ({
  type: 'SET_SPECIFIC_USER_POSTS',
  posts
});

export const startSetSpecificUserPosts = id => (dispatch) => {
  dispatch(readingsIsLoading(true));
  return db
    .collection('posts')
    .where('authorId', '==', id)
    .get()
    .then((querySnapshot) => {
      const posts = [];
      querySnapshot.forEach((doc) => {
        posts.push({
          id: doc.id,
          ...doc.data()
        });
      });
      posts.sort((a, b) => a.createdAt - b.createdAt);
      return posts;
    })
    .then(posts => dispatch(setSpecificUserPosts(posts)))
    .then(() => dispatch(startSetAuthorFromUserId(id)))
    .then(() => dispatch(readingsIsLoading(false)))
    .catch(() => dispatch(readingssHasErrored(true)));
};

// manage reading a single post.
// make it current.
export const setOnePost = post => ({
  type: 'SET_ONE_POST',
  post
});

export const startSetOnePost = id => (dispatch, getState) => {
  dispatch(readingsIsLoading(true));
  return db
    .collection('posts')
    .doc(id)
    .get()
    .then(doc =>
      dispatch(setOnePost({
        id: doc.id,
        ...doc.data()
      })))
    .then(ref => dispatch(startSetAuthorFromUserId(ref.post.authorId)))
    .then(() => dispatch(readingsIsLoading(false)))
    .catch(() => dispatch(readingssHasErrored(true)));
};
