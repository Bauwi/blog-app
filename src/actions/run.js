import database from '../firebase/firebase';
import { db } from '../firebase/firebase';
import * as firebase from 'firebase';

// manage CRUD of run
export const addPostToRun = post => ({
  type: 'ADD_POST_TO_RUN',
  post
});

export const startAddPostToRun = post => (dispatch, getState) => {
  const { uid } = getState().auth;
  db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .add(post)
    .then((doc) => {
      dispatch(addPostToRun({
        DBid: doc.id,
        ...post
      }));
    });
};

export const removePostToRun = id => ({
  type: 'REMOVE_POST_TO_RUN',
  id
});

export const startRemovePostToRun = (id, DBid) => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .doc(`${DBid}`)
    .delete()
    .then(dispatch(removePostToRun(id)));
};

// set Run posts. This action is dispatched in the header component
// as it needs to be available for the run dropdown

export const runsHasErrored = bool => ({
  type: 'RUN_HAS_ERRORED',
  hasErrored: bool
});

export const runIsLoading = bool => ({
  type: 'RUN_IS_LOADING',
  isLoading: bool
});

export const setRunPosts = posts => ({
  type: 'SET_RUN_POSTS',
  posts
});

export const startSetRunPosts = resume => (dispatch, getState) => {
  dispatch(runIsLoading(true));
  const { uid } = getState().auth;
  const run = [];
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .get()
    .then((querySnapshot) => {
      querySnapshot.forEach((doc) => {
        run.push({
          DBid: doc.id,
          ...doc.data()
        });
      });
      return run;
    })
    .then(run => dispatch(setRunPosts(run)))
    .then(() => {
      const firstUnreadIndex = run.findIndex(post => post.state === 'unread');
      if (firstUnreadIndex === -1) {
        return dispatch(setCurrentPostRun(run[0].content.id));
      }
      return dispatch(setCurrentPostRun(run[firstUnreadIndex].content.id));
    })
    .then(() => dispatch(runIsLoading(false)))
    .catch(() => dispatch(runsHasErrored(true)));
};

// change the status of the past from read to unread
export const updateRunPostToAlreadyRead = id => ({
  type: 'UPDATE_RUN_POST_TO_ALREADY_READ',
  id
});

export const startUpdateRunPostToAlreadyRead = (id, DBid) => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .doc(`${DBid}`)
    .update({ state: 'read' })
    .then(() => {
      dispatch(updateRunPostToAlreadyRead(id));
    });
};

// this one is not used anymore. But it could be used to mark some popst as unread yet
export const updateRunPostToNotAlreadyRead = id => ({
  type: 'UPDATE_RUN_POST_TO_NOT_ALREADY_READ',
  id
});

// Set which post should actually be current. Run navigation depends on this.
// instead of routing, component is updated with currentPost data.
// easier way to deal with "Resume" to provide the user the first unread post in its run.
export const setCurrentPostRun = id => ({
  type: 'SET_CURRENT_POST_RUN',
  id
});

export const startSetCurrentPostRun = id => (dispatch) => {
  dispatch(runIsLoading(true));
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve();
    }, 10);
  })
    .then(() => dispatch(setCurrentPostRun(id)))
    .then(() => dispatch(runIsLoading(false)))
    .catch(() => dispatch(runsHasErrored(true)));
};

export const resetRun = () => ({
  type: 'RESET_RUN'
});

// remove all posts in tun
export const startResetRun = () => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .get()
    .then((querySnapshot) => {
      querySnapshot.forEach((doc) => {
        db
          .collection('runs')
          .doc(uid)
          .collection('UserRuns')
          .doc(doc.ref.id)
          .delete();
      });
    })
    .then(() => dispatch(resetRun()));
};

// Remove already read posts in run
export const cleanRun = () => ({
  type: 'CLEAN_RUN'
});

export const startCleanRun = alreadyRead => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .get()
    .then((querySnapShot) => {
      querySnapShot.forEach((doc) => {
        if (alreadyRead.includes(doc.ref.id)) {
          db
            .collection('runs')
            .doc(uid)
            .collection('UserRuns')
            .doc(doc.ref.id)
            .delete();
        }
      });
    })
    .then(dispatch(cleanRun()));
};
