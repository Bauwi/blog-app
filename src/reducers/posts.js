const postsReducerDefaultState = { posts: [] };

const postsReducer = (state = postsReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_POST':
      return {
        ...state,
        posts: [...state.posts, action.post]
      };

    case 'REMOVE_POST':
      return { ...state, posts: state.posts.filter(post => post.id !== action.id) };
    case 'EDIT_POST':
      return {
        ...state,
        posts: state.posts.map((post) => {
          if (post.id === action.id) {
            return {
              ...post,
              ...action.updates
            };
          }
          return post;
        })
      };
    case 'SET_POSTS':
      return { ...state, posts: [...state.posts, ...action.posts] };
    case 'POSTS_IS_LOADING':
      return { ...state, isLoading: action.isLoading };
    case 'POSTS_HAS_ERRORED':
      return { ...state, hasErrored: action.hasErrored };
    case 'UP_POST_STAR':
      return {
        ...state,
        posts: state.posts.map((post) => {
          if (post.id === action.id) {
            return {
              ...post,
              stars: action.stars
            };
          }
          return post;
        })
      };

    default:
      return state;
  }
};

export default postsReducer;
