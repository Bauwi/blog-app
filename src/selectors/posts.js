const getVisiblePosts = (posts, { text, searchBy }) =>
  posts.filter((post) => {
    if (searchBy === 'title') {
      return post.title.toLowerCase().includes(text.toLowerCase());
    } else if (searchBy === 'keyword') {
      return post.keywords.toLowerCase().includes(text.toLowerCase());
    } else if (searchBy === 'author') {
      return post.author.toLowerCase().includes(text.toLowerCase());
    }
  });
export default getVisiblePosts;
