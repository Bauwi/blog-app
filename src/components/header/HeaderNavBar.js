import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { NavLink } from 'react-router-dom';
import { connect } from 'react-redux';
import { Badge } from 'antd';

import RunListDropdown from './RunListDropdown';
import LoginModal from './LoginModal';

import { startLogout } from '../../actions/auth';
import { startSetRunPosts } from '../../actions/run';

export class HeaderNavBar extends Component {
  componentDidMount() {
    if (this.props.isAuthenticated) {
      this.props.startSetRunPosts();
    }
  }

  render() {
    const { isAuthenticated, runCount = 0, startLogout } = this.props;
    if (isAuthenticated) {
      return (
        <nav className="header__navbar">
          <NavLink to="/">
            <i className="fa fa-home" />
          </NavLink>
          <NavLink to="/dashboard">
            <i className="fa fa-pencil" />
          </NavLink>
          <nav to="/run">
            <Badge className="header__navbar__run__badge" showZero count={runCount} />
            <RunListDropdown />
          </nav>
          <NavLink to="/preferences">
            <i className="fa fa-cog" />
          </NavLink>
          <button className="button button--navbar" onClick={startLogout}>
            <i className="fa fa-sign-out" />
          </button>
        </nav>
      );
    }
    return (
      <nav className="header__navbar">
        <NavLink to="/">
          <i className="fa fa-home" />
        </NavLink>
        <LoginModal />
      </nav>
    );
  }
}

HeaderNavBar.propTypes = {
  isAuthenticated: PropTypes.bool.isRequired,
  runCount: PropTypes.number,
  startLogout: PropTypes.func.isRequired,
  startSetRunPosts: PropTypes.func.isRequired
};

HeaderNavBar.defaultProps = { runCount: 0 };

const mapDispatchToProps = dispatch => ({
  startLogout: () => dispatch(startLogout()),
  startSetRunPosts: () => dispatch(startSetRunPosts())
});

const mapStateToProps = state => ({
  isAuthenticated: !!state.auth.uid,
  runCount: state.run.posts ? state.run.posts.length : 0
});

export default connect(mapStateToProps, mapDispatchToProps)(HeaderNavBar);
