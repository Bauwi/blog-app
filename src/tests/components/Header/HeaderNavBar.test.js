import React from 'react';
import { shallow } from 'enzyme';
import { HeaderNavBar } from '../../../components/header/HeaderNavBar';
import { startLogout } from '../../../actions/auth';
import { startSetRunPosts } from '../../../actions/posts';

test('should render HeaderNavBar correctly', () => {
  const wrapper = shallow(<HeaderNavBar startLogout={() => {}} startSetRunPosts={() => {}} isAuthenticated />);
  expect(wrapper).toMatchSnapshot();
});

test('should call startLogout on button click', () => {
  const startLogout = jest.fn();
  const wrapper = shallow(<HeaderNavBar startLogout={startLogout} startSetRunPosts={() => {}} isAuthenticated />);
  wrapper.find('button').simulate('click');

  expect(startLogout).toHaveBeenCalled();
});

test('should call startSetRunPosts on mount', () => {
  const startSetRunPosts = jest.fn();
  const wrapper = shallow(<HeaderNavBar startLogout={() => {}} startSetRunPosts={startSetRunPosts} isAuthenticated />);

  expect(startSetRunPosts).toHaveBeenCalled();
});

test('should not call startSetRunPost if not authenticated', () => {
  const startSetRunPosts = jest.fn();
  const wrapper = shallow(<HeaderNavBar
    startLogout={() => {}}
    startSetRunPosts={startSetRunPosts}
    isAuthenticated={false}
  />);

  expect(startSetRunPosts).not.toHaveBeenCalled();
});
