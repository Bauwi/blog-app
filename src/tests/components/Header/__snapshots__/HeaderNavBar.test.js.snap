// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render HeaderNavBar correctly 1`] = `
<nav
  className="header__navbar"
>
  <NavLink
    activeClassName="active"
    ariaCurrent="true"
    to="/"
  >
    <i
      className="fa fa-home"
    />
  </NavLink>
  <NavLink
    activeClassName="active"
    ariaCurrent="true"
    to="/dashboard"
  >
    <i
      className="fa fa-pencil"
    />
  </NavLink>
  <nav
    to="/run"
  >
    <Badge
      className="header__navbar__run__badge"
      count={0}
      dot={false}
      overflowCount={99}
      prefixCls="ant-badge"
      scrollNumberPrefixCls="ant-scroll-number"
      showZero={true}
    />
    <withRouter(Connect(RunListDropdown)) />
  </nav>
  <NavLink
    activeClassName="active"
    ariaCurrent="true"
    to="/preferences"
  >
    <i
      className="fa fa-cog"
    />
  </NavLink>
  <button
    className="button button--navbar"
    onClick={[Function]}
  >
    <i
      className="fa fa-sign-out"
    />
  </button>
</nav>
`;
